### PR TITLE
[AdamW] Fix complex x amsgrad support

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104990
* #104989

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Ho good catch, thanks!

File: torch/optim/adamw.py [hunk 1]
```diff
@@ -424,15 +424,19 @@ def _single_tensor_adamw(
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
                 if differentiable:
-                    max_exp_avg_sqs_i = max_exp_avg_sqs[i].clone()
+                    max_exp_avg_sq = max_exp_avg_sq.clone()
+                torch.maximum(max_exp_avg_sq, exp_avg_sq, out=max_exp_avg_sq)
```
Comment (albanD) [inline torch/optim/adamw.py:428]: out= variants are NOT differentiable! You won't be able to do that

File: torch/optim/adamw.py [hunk 2]
```diff
@@ -424,15 +424,19 @@ def _single_tensor_adamw(
...
-                    max_exp_avg_sqs_i = max_exp_avg_sqs[i]
-                max_exp_avg_sqs[i].copy_(torch.maximum(max_exp_avg_sqs_i, exp_avg_sq))
+                    max_exp_avg_sqs[i] = max_exp_avg_sq
+
                 # Uses the max. for normalizing running avg. of gradient
                 # Folds in (admittedly ugly) 1-elem step_size math here to avoid extra param-set-sized read+write
                 # (can't fold it into addcdiv_ below because addcdiv_ requires value is a Number, not a Tensor)
                 denom = (
-                    max_exp_avg_sqs[i].sqrt() / (bias_correction2_sqrt * step_size_neg)
+                    max_exp_avg_sq.sqrt() / (bias_correction2_sqrt * step_size_neg)
```
Comment (albanD) [inline torch/optim/adamw.py:439]: You want the complex version of this variable here?

Comment (janeyx99) [reply torch/optim/adamw.py:439]: I want the one that’s been view_as_real’d

**Pull request closed without merging**